### PR TITLE
drivers: bluetooth: hci: Support reading long events in ST SPI v1

### PR DESCRIPTION
**Pull request by HoZHel**
Provide support for reading events longer than 127 bytes in ST SPI protocol v1 by splitting them into several SPI transactions.

### CODE REVIEW

**Review by jori-nordic — DISMISSED** [review]

**Review by msmttchr — APPROVED** [review]
LGTM

File: drivers/bluetooth/hci/hci_spi_st.c [hunk 1]
```diff
@@ -334,13 +334,27 @@ static int bt_spi_bluenrg_setup(const struct bt_hci_setup_params *params)
...
-	struct net_buf *buf;
-	int len;
+	static int len;
+	struct net_buf *buf = *bufp;
+	int ret = 0;
+
+#if DT_HAS_COMPAT_STATUS_OKAY(st_hci_spi_v1)
+	if (buf) {
+		/* Buffer already allocated, waiting to complete event reception */
+		net_buf_add_mem(buf, msg, MIN(size, len-buf->len));
```
Comment (jhedberg) [inline drivers/bluetooth/hci/hci_spi_st.c:349]: Coding style (and consistency with yourself): `len - buf->len`

File: drivers/bluetooth/hci/hci_spi_st.c [hunk 2]
```diff
@@ -334,13 +334,27 @@ static int bt_spi_bluenrg_setup(const struct bt_hci_setup_params *params)
...
+		if ((buf)->len >= len) {
```
Comment (jhedberg) [inline drivers/bluetooth/hci/hci_spi_st.c:350]: Unnecessary `()` around `buf`

File: drivers/bluetooth/hci/hci_spi_st.c [hunk 3]
```diff
@@ -375,17 +389,22 @@ static struct net_buf *bt_spi_rx_buf_construct(uint8_t *msg)
...
 
 		len = sizeof(struct bt_hci_evt_hdr) + msg[EVT_HEADER_SIZE];
 		if (len > net_buf_tailroom(buf)) {
 			LOG_ERR("Event too long: %d", len);
 			net_buf_unref(buf);
-			return NULL;
+			return -ENOMEM;
 		}
-		net_buf_add_mem(buf, &msg[1], len);
+		net_buf_add_mem(buf, &msg[1], size-1);
```
Comment (jhedberg) [inline drivers/bluetooth/hci/hci_spi_st.c:402]: `size - 1` (missing spaces around `-`)

File: drivers/bluetooth/hci/hci_spi_st.c [hunk 4]
```diff
@@ -334,13 +334,27 @@ static int bt_spi_bluenrg_setup(const struct bt_hci_setup_params *params)
...
 
-static struct net_buf *bt_spi_rx_buf_construct(uint8_t *msg)
+static int bt_spi_rx_buf_construct(uint8_t *msg, struct net_buf **bufp, uint16_t size)
 {
 	bool discardable = false;
 	k_timeout_t timeout = K_FOREVER;
 	struct bt_hci_acl_hdr acl_hdr;
-	struct net_buf *buf;
-	int len;
+	static int len;
```
Comment (jhedberg) [inline drivers/bluetooth/hci/hci_spi_st.c:342]: I was scratching my head around some of the new code, until I realised you're transforming `len` into a static variable. Maybe add a code comment here or rename it to something more descriptive (that gives a better hint at its persistency)? Alternatively, could the same be solved by moving it out into the loop of the calling function (like you're doing with `bufp`)?

**Review by jhedberg — CHANGES_REQUESTED** [review]

File: drivers/bluetooth/hci/hci_spi_st.c [hunk 5]
```diff
@@ -375,17 +389,22 @@ static struct net_buf *bt_spi_rx_buf_construct(uint8_t *msg)
...
 		if (len > net_buf_tailroom(buf)) {
 			LOG_ERR("Event too long: %d", len);
 			net_buf_unref(buf);
-			return NULL;
+			return -ENOMEM;
 		}
-		net_buf_add_mem(buf, &msg[1], len);
+		net_buf_add_mem(buf, &msg[1], size-1);
+#if DT_HAS_COMPAT_STATUS_OKAY(st_hci_spi_v1)
+		if ((size - 1) < len) {
```
[the remark this answers is not in the document]
Comment (HoZHel) [reply drivers/bluetooth/hci/hci_spi_st.c]: According to this scope,
https://github.com/zephyrproject-rtos/zephyr/blob/f9f44b6dcdd4b80765df8922e8d12fdb9213e13a/drivers/bluetooth/hci/hci_spi_st.c#L441-L446
`size == 0` is handled properly, so it doesn't reach this part.

A comment has been added to show the purpose of `size - 1`.

File: drivers/bluetooth/hci/hci_spi_st.c [hunk 6]
```diff
@@ -349,7 +363,7 @@ static struct net_buf *bt_spi_rx_buf_construct(uint8_t *msg)
 		struct bt_hci_evt_hdr *evt2 = (struct bt_hci_evt_hdr *) (msg + 1);
 
 		if (evt->len > 0xff) {
```
[the remark this answers is not in the document]
Comment (HoZHel) [reply drivers/bluetooth/hci/hci_spi_st.c]: Thanks for noticing. Fixed.

File: drivers/bluetooth/hci/hci_spi_st.c (hunk 1 above)
Comment (HoZHel) [reply drivers/bluetooth/hci/hci_spi_st.c]: Fixed.

File: drivers/bluetooth/hci/hci_spi_st.c (hunk 2 above)
Comment (HoZHel) [reply drivers/bluetooth/hci/hci_spi_st.c]: Fixed.

File: drivers/bluetooth/hci/hci_spi_st.c (hunk 3 above)
Comment (HoZHel) [reply drivers/bluetooth/hci/hci_spi_st.c]: Done.

File: drivers/bluetooth/hci/hci_spi_st.c (hunk 4 above)
Comment (HoZHel) [reply drivers/bluetooth/hci/hci_spi_st.c]: I've added a descriptive comment for `len`.
Alternative solution requires more changes on the code.

**Review by jhedberg — APPROVED** [review]

**Review by erwango — APPROVED** [review]

**Pull request merged**